### PR TITLE
Event poc [Don't Merge!]

### LINKED ISSUE #580: IDE act: Eventing prototype
A small proof-of-concept to catch events needed for the workflow of the one presented in #561 and #562 or similar. A POC should show that those events can be caught and the data related to them (like, contents of the saved file) can be accessed.

The design can be very crude. No workflow needs to be implemented.

### PR DESCRIPTION
**Pull request by stellatsv**
# Description of the PR
Closes #580. This is a prototype of the different listeners that could be used in implementing the Assessable IDE activities. 
There are listeners registered for the following things:
- Save events (triggered when saving if there are changes). The file's text can also be retrieved.
- File rename events
- Existing IntelliJ Actions (like run, debug, save, etc.) and the plugin's Actions
- File opened events
- Renaming events (through the Rename action of IntelliJ), from where we can access the new name and also the element's old name. Also, we can see if the element that was renamed was a PsiField, PsiMethod, PsiClass.

In order to check the functionality go to A+ > Start PoC Tutorial (this Action registers the listeners).

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [ ] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [ ] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [ ] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/PocAction.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    public void beforeActionPerformed(@NotNull AnAction action, @NotNull DataContext dataContext,
+                                      @NotNull AnActionEvent event) {
+      System.out.println("Action: " + action.getTemplateText()
+              + " ActionEvent: " + event.getPlace());
+    }
+  }
+
+  private class RefactoringListener implements RefactoringEventListener {
+    @RequiresReadLock
+    public synchronized void registerListeners() {
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/PocAction.java:73]: `APlusActionListener;;registerListener` doesn't have `@RequiresReadLock`? Is that intentional that there's a difference (they're, after all, identical except for topic). If not, do you know which one is the correct one?

**Review by OlliKiljunen — COMMENTED** [review]
Perfect! 🤩 Just what we needed. Very well done!

I like the structured way this PoC is coded because it makes it easier to think of what the actual design of the IDE-act "engine" could be like. 👍

(This is a comment-only-without-explicit-approval review because this PR is do-not-merge.)

**Review by jaakkonarhi — COMMENTED** [review]
Everything seems to work well

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/PocAction.java (hunk 1 above)
Comment (stellatsv) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/actions/PocAction.java:74]: Not intentional, I missed adding the `@RequiresReadLock` to `APlusActionListener`.

**Pull request closed without merging**